src/helpers/DecimalBigNumber.ts

```diff
@@ -0,0 +1,60 @@
+import { BigNumber } from "@ethersproject/bignumber";
+import { formatUnits, parseUnits } from "@ethersproject/units";
+
+import { formatNumber } from ".";
+
+export class DecimalBigNumber {
+  private _decimals: number;
+  private _number: BigNumber;
+
+  constructor(number: BigNumber | string, decimals: number) {
+    this._decimals = decimals;
+    this._number = typeof number === "string" ? parseUnits(number, decimals) : number;
+  }
+
+  /**
+   * Use when performing accurate calculations with
+   * the number where precision is important.
+   */
+  public toBigNumber(): BigNumber {
+    return this._number;
+  }
+
+  /**
+   * Used to display the value accurately to the user
+   */
+  public toAccurateString(): string {
+    return formatUnits(this._number, this._decimals);
+  }
+
+  /**
+   * Use when performing approximate calculations with
+   * the number where precision __is not__ important.
+   */
+  public toApproxNumber(): number {
+    return parseFloat(this.toAccurateString());
+  }
+
+  /**
+   * Used to display a formatted approximate value to the user
+   * @param precision The number of decimal places to show
+   */
+  public toFormattedString(precision = this._decimals): string {
+    const number = this.toApproxNumber();
+
+    return formatNumber(number, precision);
+  }
+
+  /**
+   * Adds two numbers
+   */
+  public add(value: DecimalBigNumber) {
+    const difference = value._decimals - this._decimals;
+
+    if (difference === 0) return new DecimalBigNumber(this._number.add(value._number), this._decimals);
+
+    return difference > 0
+      ? new DecimalBigNumber(this._number.mul(10 ** Math.abs(difference)).add(value._number), value._decimals)
+      : new DecimalBigNumber(this._number.add(value._number.mul(10 ** Math.abs(difference))), this._decimals);
+  }
+}
```

Comment:
interesting so the intent is to normalize math between two different Big Numbers with different decimal values...

Reply:
Yep. Because BigNumber has no concept of decimals and stores everything as integers, if we don't normalize the math doesn't work as we would expect. So adding two numbers with different precision - 9.1 and 9.12 - is performed under the hood as 91 + 912 = 1003, but by normalizing: 910 + 912 = 1822 (returned as a `DecimalBigNumber` to two decimal places)

Reply:
yea that's much better than doing it by hand within all the thunks... :|